File: days/src/main/java/aoc2024j/day12/Day12.java

package aoc2024j.day12;

import utils.CharGrid;
import utils.IO;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Gatherers;
import java.util.stream.Stream;

public class Day12 {

    static class PlantMap extends CharGrid {

        public PlantMap(List<String> data) {
            super(data, false);
        }

        Regions findRegions() {
            var regions = new Regions();
            var assigned = new boolean[height][width];
            for (int y = 0; y < height; y++) {
                for (int x = 0; x < width; x++) {
                    if (assigned[y][x]) continue;
                    var plant = points[y][x];
                    Region region = expandRegion(x, y, assigned);
                    regions.addRegion(plant, region);
                }
            }
            return regions;
        }

        private Region expandRegion(int x, int y, boolean[][] assigned) {
            var region = new Region(new GardenPlot(x, y));
            var queue = new LinkedList<GardenPlot>();
            queue.add(new GardenPlot(x, y));
            assigned[y][x] = true;
            while (!queue.isEmpty()) {
                var plot = queue.poll();
                for (var neighbor : neighbors(plot)) {
                    if (!assigned[neighbor.y][neighbor.x] && points[neighbor.y][neighbor.x] == points[y][x]) {
                        assigned[neighbor.y][neighbor.x] = true;
                        region.add(neighbor);
                        queue.add(neighbor);
                    }
                }
            }
            return region;
        }

        private List<GardenPlot> neighbors(GardenPlot plot) {
            return Stream.of(
                    new GardenPlot(plot.x - 1, plot.y),
                    new GardenPlot(plot.x + 1, plot.y),
                    new GardenPlot(plot.x, plot.y - 1),
                    new GardenPlot(plot.x, plot.y + 1)
            ).filter(p -> p.x >= 0 && p.x < width && p.y >= 0 && p.y < height).toList();
        }

    }

    static class Regions {
        private final Map<Character, List<Region>> regions = new HashMap<>();

        public void addRegion(char plant, Region region) {
            regions.computeIfAbsent(plant, _ -> new ArrayList<>()).add(region);
        }

        public long price1() {
            return regions.values().stream()
                    .flatMap(List::stream)
                    .mapToLong(Region::price1)
                    .sum();
        }

        public long price2() {
            return regions.values().stream()
                    .flatMap(List::stream)
                    .mapToLong(Region::price2)
                    .sum();
        }
    }

    static class Region {
        final List<GardenPlot> plots = new ArrayList<>();
        Map<Orientation, Map<Integer, List<Wall>>> cachedWalls = null;

        private Map<Orientation, Map<Integer, List<Wall>>> walls() {
            if (cachedWalls == null) {
                cachedWalls = GardenPlot.generateWalls(plots);
            }
            return cachedWalls;
        }

        public Region(GardenPlot plot) {
            plots.add(plot);
        }

        public void add(GardenPlot plot) {
            plots.add(plot);
            cachedWalls = null;
        }

        public long price1() {
            return area() * perimeter();
        }

        public long price2() {
            return area() * sides();
        }

        private long area() {
            return plots.size();
        }

        private long perimeter() {
            return walls().values().stream()
                    .flatMap(m -> m.values().stream())
                    .mapToLong(List::size)
                    .sum();
        }

        private long sides() {
            return walls().values().stream()
                    .flatMap(m -> m.values().stream())
                    .mapToLong(Wall::countWalls)
                    .sum();
        }

    }

    record GardenPlot(int x, int y) {

        private static Map<Orientation, Map<Integer, List<Wall>>> generateWalls(List<GardenPlot> plots) {
            return plots.stream()
                    .flatMap(GardenPlot::walls)
                    .collect(Collectors.groupingBy(
                            s -> s.direction().orientation(),
                            Collectors
                                    .groupingBy(
                                            Wall::ref,
                                            Collectors.collectingAndThen(
                                                    Collectors.toList(),
                                                    walls -> {
                                                        walls.sort(Comparator.comparingInt(Wall::coordinate));
                                                        return Wall.filterWalls(walls);
                                                    }
                                            ))));
        }

        private Stream<Wall> walls() {
            return Stream.of(
                    new Wall(y, x, Direction.UP),
                    new Wall(y + 1, x, Direction.DOWN),
                    new Wall(x, y, Direction.LEFT),
                    new Wall(x + 1, y, Direction.RIGHT)
            );
        }
    }

    enum Orientation {
        HORIZONTAL, VERTICAL
    }

    enum Direction {
        UP, DOWN, LEFT, RIGHT;

        Orientation orientation() {
            return this == UP || this == DOWN ? Orientation.HORIZONTAL : Orientation.VERTICAL;
        }

        static Stream<Direction> all() {
            return Stream.of(UP, DOWN, LEFT, RIGHT);
        }
    }

    record Wall(int ref, int coordinate, Direction direction) {

        static long countWalls(List<Wall> walls) {
            return
                    walls.stream().collect(
                            Collectors.collectingAndThen(
                                    Collectors.groupingBy(Wall::direction,
                                            Collectors.mapping(Wall::coordinate,
                                                    Collectors.toList())),
                                    coordinatesByDirection -> Direction.all()
                                            .mapToLong(d -> countGaps(coordinatesByDirection.getOrDefault(d, List.of())))
                                            .sum()));
        }

        private static long countGaps(List<Integer> coordinates) {
            if (coordinates.isEmpty()) return 0;
            long counter = 1;
            var it = coordinates.iterator();
            var current = it.next();
            while (it.hasNext()) {
                var next = it.next();
                if (next - current > 1) {
                    counter++;
                }
                current = next;
            }
            return counter;
        }

        private static List<Wall> filterWalls(List<Wall> candidateWalls) {
            if (candidateWalls.size() < 2) {
                return candidateWalls;
            }
            var purged = candidateWalls.stream()
                    .gather(Gatherers.windowSliding(2))
                    .filter(l -> l.get(0).coordinate() == l.get(1).coordinate())
                    .map(l -> l.get(0).coordinate())
                    .collect(Collectors.toSet());
            return candidateWalls.stream()
                    .filter(wall -> !purged.contains(wall.coordinate()))
                    .collect(Collectors.toList());
        }
    }

    long part1(List<String> data) {
        var map = new PlantMap(data);
        var regions = map.findRegions();
        return regions.price1();
    }

    long part2(List<String> data) {
        var map = new PlantMap(data);
        var regions = map.findRegions();
        return regions.price2();
    }

    public static void main() {
        var day12 = new Day12();
        var data = IO.getResourceAsList("aoc2024/day12.txt");
        var part1 = day12.part1(data);
        System.out.println("part1 = " + part1);
        var part2 = day12.part2(data);
        System.out.println("part2 = " + part2);
    }
}

File: days/src/main/scala/aoc2024s/day12/Day12.scala

package aoc2024s.day12;

import utils.IO

import scala.annotation.targetName
import scala.collection.mutable
import scala.jdk.CollectionConverters.*

object Day12 {

  class PlantMap(data: List[String]) {
    val points: Vector[String] = data.toVector
    val height: Int = points.size
    val width: Int = points(0).length

    def findRegions: Regions = {
      val regions = new Regions(this)
      val assigned = Array.fill(height, width)(false)
      for {
        y <- 0 until height
        x <- 0 until width
        if !assigned(y)(x)
      } {
        val plant = points(y)(x)
        val region = expandRegion(x, y, assigned)
        regions.addRegion(plant, region)
      }
      regions
    }

    def expandRegion(x: Int, y: Int, assigned: Array[Array[Boolean]]): Region = {
      val region = new Region(GardenPlot(x, y))
      val queue = mutable.Queue(GardenPlot(x, y))
      assigned(y)(x) = true
      while (queue.nonEmpty) {
        val plot = queue.dequeue()
        for {
          neighbor <- plot.neighbors
          if neighbor.x >= 0 && neighbor.x < width && neighbor.y >= 0 && neighbor.y < height
          if !assigned(neighbor.y)(neighbor.x)
          if points(neighbor.y)(neighbor.x) == points(y)(x)
        } {
          assigned(neighbor.y)(neighbor.x) = true
          region.add(neighbor)
          queue.enqueue(neighbor)
        }
      }
      region
    }

    def get(pos: GardenPlot): Int =
      if (pos.x < 0 || pos.x >= width || pos.y < 0 || pos.y >= height) -1
      else points(pos.y)(pos.x)

    def isCorner(direction1: Direction, direction2: Direction)(pos: GardenPlot): Boolean =
      val value = get(pos)
      val side1 = get(direction1(pos))
      val side2 = get(direction2(pos))
      val corner = get((direction1 + direction2)(pos))
      corner != value && side1 == value && side2 == value
        || corner != value && side1 != value && side2 != value
        || corner == value && side1 != value && side2 != value

  }

  class Regions(plantMap: PlantMap) {
    val regions: mutable.HashMap[Char, List[Region]] = mutable.HashMap.empty

    def addRegion(plant: Char, region: Region): Unit = {
      regions.get(plant) match {
        case Some(other) => regions.put(plant, region :: other)
        case None => regions.put(plant, List(region))
      }
    }

    def price1: Long =
      regions.valuesIterator.flatten.map(_.price1).sum

    def price2: Long =
      regions.valuesIterator.flatten.map(_.price2(plantMap)).sum
  }

  class Region {
    val plots: mutable.ArrayBuffer[GardenPlot] = mutable.ArrayBuffer.empty

    def this(plot: GardenPlot) = {
      this()
      plots += plot
    }

    def add(plot: GardenPlot): Unit = {
      plots += plot
    }

    def area: Long = plots.size

    def perimeter: Long =
      plots.flatMap(_.neighbors).filterNot(plots.contains).size

    def sides(plantMap: PlantMap): Long =
      (for {
        vertical <- List(Direction.N, Direction.S)
        horizontal <- List(Direction.W, Direction.E)
        isCorner = plantMap.isCorner(vertical, horizontal)
        plot <- plots
        if isCorner(plot)
      } yield ()).size

    def price1: Long = area * perimeter

    def price2(plantMap: PlantMap): Long = area * sides(plantMap)
  }

  enum Direction(val dx: Int, val dy: Int) extends (GardenPlot => GardenPlot) {
    case N extends Direction(0, -1)
    case S extends Direction(0, 1)
    case E extends Direction(1, 0)
    case W extends Direction(-1, 0)

    def apply(plot: GardenPlot): GardenPlot =
      GardenPlot(plot.x + dx, plot.y + dy)

    @targetName("andThen")
    def +(other: Direction): GardenPlot => GardenPlot = compose(other)
  }

  case class GardenPlot(x: Int, y: Int) {
    def neighbors: Iterator[GardenPlot] =
      Iterator(
        GardenPlot(x - 1, y),
        GardenPlot(x + 1, y),
        GardenPlot(x, y - 1),
        GardenPlot(x, y + 1)
      )
  }

  def part1(data: List[String]): Long = {
    val map = new PlantMap(data)
    val regions = map.findRegions
    regions.price1
  }

  def part2(data: List[String]): Long = {
    val map = new PlantMap(data)
    val regions = map.findRegions
    regions.price2
  }

  @main def main12(): Unit = {
    val data = IO.getResourceAsList("aoc2024/day12.txt").asScala.toList;
    val part1 = Day12.part1(data);
    println(s"part1 = $part1");
    val part2 = Day12.part2(data);
    println(s"part2 = $part2");
  }
}